### PR TITLE
[TE] Rootcause analysis entities and utilities

### PR DESCRIPTION
**Pull request by apucher**
* Entities

    * MetricEntity

    * DimensionEntity

    * EventEntity

    * ServiceEntity

    * TimeRangeEntity

* EntityUtils

* builds on #1361

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/DimensionEntity.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * DimensionEntity represents a data dimension (a cut) across multiple metrics. It is identified
+ * by a key-value pair. Note, that dimension names may require standardization across different
+ * metrics.
+ */
+public class DimensionEntity extends Entity {
+  public static final EntityType TYPE = new EntityType("thirdeye:dimension:");
+
+  public static DimensionEntity fromDimension(double score, String name, String value) {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/DimensionEntity.java]: move methods after the constructor. keep variables at the top of the class

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityType.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  public static String extractPrefix(Entity e) {
+    return extractPrefix(e.getUrn());
+  }
+
+  public String getPrefix() {
+    return prefix;
+  }
+
+  public EntityType(String prefix) {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityType.java]: why do we need this constructor? Shouldn't this class be abstract?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * EventEntity represents an individual event. It holds meta-data referencing ThirdEye's internal
+ * database. The URN namespace is defined as 'thirdeye:event:{type}:{name}:{start}'.
+ */
+public class EventEntity extends Entity {
+  public static final EntityType TYPE = new EntityType("thirdeye:event:");
+
+  public static EventEntity fromDTO(double score, EventDTO dto) {
+    String urn = TYPE.formatURN(dto.getEventType(), dto.getName(), dto.getStartTime());
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java]: any reason to not have endTime in the event entity urn?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  final EventDTO dto;
+
+  protected EventEntity(String urn, double score, EventDTO dto) {
+    super(urn, score);
+    this.dto = dto;
+  }
+
+  public EventDTO getDto() {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java]: Dto -> DTO

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityType.java (hunk 2 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityType.java]: EntityType serves the same purpose as an Enum - with the difference that it's extensible by other projects importing ThirdEye as a library.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java (hunk 3 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java]: we require a reference to an EventDTO anyways

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java (hunk 4 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EventEntity.java]: done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/DimensionEntity.java (hunk 1 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/DimensionEntity.java]: done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/DimensionEntity.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import com.linkedin.thirdeye.rootcause.Entity;
+
+
+/**
+ * DimensionEntity represents a data dimension (a cut) across multiple metrics. It is identified
+ * by a key-value pair. Note, that dimension names may require standardization across different
+ * metrics.
+ */
+public class DimensionEntity extends Entity {
+  public static final EntityType TYPE = new EntityType("thirdeye:dimension:");
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/DimensionEntity.java]: see comment about EntityType.java below

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityType.java (hunk 2 above)
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityType.java]: Got it, I confused this with Entity

**Comment by kishoreg** [comment]
LGTM

**Pull request merged**